cdap-client/src/main/java/co/cask/cdap/client/MetricsClient.java

```diff
@@ -59,6 +63,62 @@ public MetricsClient(ClientConfig config) {
   }
 
   /**
+   * Searches for metrics matching the given tags.
+   *
+   * @param tags the tags to match
+   * @return the metrics matching the given tags
+   * @throws IOException if a network error occurred
+   * @throws UnauthorizedException if the request is not authorized successfully in the gateway server
+   */
+  public List<MetricTagValue> search(List<String> tags)
+    throws IOException, UnauthorizedException {
+
+    List<String> queryParts = Lists.newArrayList();
+    queryParts.add("target=tag");
+    for (String tag : tags) {
+      queryParts.add("tag=" + tag);
+    }
+
+    URL url = config.resolveURLV3(String.format("metrics/search?%s", Joiner.on("&").join(queryParts)));
+    HttpResponse response = restClient.execute(HttpMethod.POST, url, config.getAccessToken());
+    ObjectResponse<List<MetricTagValue>> result = ObjectResponse.fromJsonBody(
+      response, new TypeToken<List<MetricTagValue>>() { }.getType());
+    return result.getResponseObject();
+  }
+
+  /**
+   * Gets the value of the given metrics.
+   *
+   * @param metrics names of the metrics
+   * @param groupBys groupBys for the request
+   * @param tags tags for the request
+   * @return values of the metrics
+   * @throws IOException if a network error occurred
+   * @throws UnauthorizedException if the request is not authorized successfully in the gateway server
+   */
+  public MetricQueryResult query(List<String> metrics, List<String> groupBys, List<String> tags)
+    throws IOException, UnauthorizedException {
+
+    List<String> queryParts = Lists.newArrayList();
+    queryParts.add("target=tag");
+    for (String metric : metrics) {
+      queryParts.add("metric=" + metric);
+    }
+    for (String groupBy : groupBys) {
+      queryParts.add("groupBy=" + groupBy);
+    }
+    for (String tag : tags) {
+      queryParts.add("tag=" + tag);
+    }
+
+    URL url = config.resolveURLV3(String.format("metrics/query?%s", Joiner.on("&").join(queryParts)));
+    HttpResponse response = restClient.execute(HttpMethod.POST, url, config.getAccessToken());
+    return ObjectResponse.fromJsonBody(response, MetricQueryResult.class).getResponseObject();
+  }
+
+  /**
+   * @deprecated As of 3.1.0, replaced by {@link #query}
```

Comment:
we will remove the context endpoints in metrics in 3.1,  as we already deprecated them in 3.0, so when we remove that, this query method would also be removed not deprecated.